### PR TITLE
CASSANDRA-17523 Reduce histogram snapshot long[] allocation overhead during speculative read and write threshold updates

### PR DESCRIPTION
**Pull request by maedhroz**

### CODE REVIEW

File: src/java/org/apache/cassandra/metrics/DecayingEstimatedHistogramReservoir.java [hunk 1]
```diff
@@ -673,6 +624,109 @@ public void dump(OutputStream output)
...
         }
+    }
+
+    private static class DecayingBucketsOnlySnapshot extends AbstractSnapshot
+    {   
+        private final long count;
+
+        public DecayingBucketsOnlySnapshot(DecayingEstimatedHistogramReservoir reservoir)
+        {
+            super (reservoir);
```
Comment (jmckenzie-dev) [inline src/java/org/apache/cassandra/metrics/DecayingEstimatedHistogramReservoir.java]: nit: extra space on method call

File: src/java/org/apache/cassandra/metrics/DecayingEstimatedHistogramReservoir.java [hunk 2]
```diff
@@ -673,6 +624,109 @@ public void dump(OutputStream output)
                 }
             }
         }
+    }
+
+    private static class DecayingBucketsOnlySnapshot extends AbstractSnapshot
```
Comment (jmckenzie-dev) [inline src/java/org/apache/cassandra/metrics/DecayingEstimatedHistogramReservoir.java]: Consider adding class level JavaDoc to briefly summarize how this differs from `EstimatedHistogramReservoirSnapshot` (also: symmetry. :) )

File: src/java/org/apache/cassandra/metrics/DecayingEstimatedHistogramReservoir.java [hunk 3]
```diff
@@ -673,6 +624,109 @@ public void dump(OutputStream output)
...
+            return Ints.saturatedCast(count);
+        }
+    }
+
+    /**
+     * Represents a snapshot of the decaying histogram.
+     *
+     * The decaying buckets are copied into a snapshot array to give a consistent view for all getters. However, the
+     * copy is made without a write-lock and so other threads may change the buckets while the array is copied,
+     * probably causign a slight skew up in the quantiles and mean values.
```
Comment (jmckenzie-dev) [inline src/java/org/apache/cassandra/metrics/DecayingEstimatedHistogramReservoir.java]: nit: spelling (may as well correct while we're here). causign -> causing

File: test/unit/org/apache/cassandra/metrics/DecayingEstimatedHistogramReservoirTest.java [hunk 4]
```diff
@@ -43,594 +49,614 @@
...
+                .forAll(offsets.flatMap(offs -> this.offsetsAndValue(offs, false, 9)))
+                .check(this::checkEstimation);
+        }
 
-        int model = findIndexModel(offsets, value);
-        int actual = DecayingEstimatedHistogramReservoir.findIndex(offsets, value);
+        //shows that the max before overflow is 238 buckets regardless of consider zeros
+        @Test
+        @Ignore
+        public void showHistorgramOffsetOverflow()
```
Comment (jmckenzie-dev) [inline test/unit/org/apache/cassandra/metrics/DecayingEstimatedHistogramReservoirTest.java]: nit: spelling. showHistorgram -> showHistogram

File: test/unit/org/apache/cassandra/metrics/DecayingEstimatedHistogramReservoirTest.java [hunk 5]
```diff
@@ -43,594 +49,614 @@
...
-        ExecutorService executors = Executors.newFixedThreadPool(nStripes * 2);
-        for (int i = 0; i < 1_000_000; i++)
+        @Parameterized.Parameter
+        public String description;
+        
+        @Parameterized.Parameter(1)
+        public Function<DecayingEstimatedHistogramReservoir, Snapshot> toSnapshot;
+
+        @Parameterized.Parameters(name="{0}")
+        public static Collection<Object[]> supliers()
```
Comment (jmckenzie-dev) [inline test/unit/org/apache/cassandra/metrics/DecayingEstimatedHistogramReservoirTest.java]: nit: spelling. supliers -> suppliers

**Review by jmckenzie-dev — APPROVED** [review]
+1 w/some optional nits you can address on merge

**Comment by maedhroz** [comment]
Committed in https://github.com/apache/cassandra/commit/af90c3634007ba9b2967267d6f8383d38cdff85f

**Pull request closed without merging**